### PR TITLE
Unicorn refactoring

### PR DESCRIPTION
**Pull request by antmat**
Main points:
  * Now API is stateless as it should be. State related to subscriptions and locks is moved to separate unicorn_scope_t class
  * Plugin code is restructured -  code is splitted to different files to make it easier to understand.
  * Functional tests

### CODE REVIEW

**Comment by 3Hren** [comment]
I'm going to review this today, stay tuned.

File: unicorn/include/cocaine/api/unicorn.hpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+           const unicorn::value_t& value,
+           bool ephemeral,
+           bool sequence) = 0;
+
+    unicorn_scope_ptr
+    create_default(writable_ptr::create result,
+                   const unicorn::path_t& path,
+                   const unicorn::value_t& value)
+    {
+        return create(std::move(result), std::move(path), std::move(value), false, false);
```
Comment (3Hren) [inline unicorn/include/cocaine/api/unicorn.hpp]: You cast const lvalue reference to mutable rvalue. Are you sure that it's exactly what you want?

File: unicorn/include/cocaine/unicorn/writable.hpp [hunk 2]
```diff
@@ -54,12 +54,12 @@ public writable_adapter_base_t<T>
...
     virtual void
     abort(const std::error_code& code, const std::string& reason) {
-        w.abort(code, reason);
-    }
-
-    virtual void
-    abort(const std::error_code& code) {
-        w.abort(code, "");
+        std::string full_reason = code.message();
+        if(!reason.empty()) {
```
Comment (3Hren) [inline unicorn/include/cocaine/unicorn/writable.hpp]: And if it is empty we just ignore the `reason` variable?

File: unicorn/src/service/unicorn.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    service_t(context, _asio, _name, args),
+    dispatch<io::unicorn_tag>(_name),
+    name(_name),
+    unicorn(api::unicorn(context, args.as_object().at("backend", "core").as_string())),
+
+
+
+    log(context.log("unicorn"))
+{
+    on<scope::subscribe>         (std::make_shared<subscribe_slot_t>         (this, unicorn, &api::unicorn_t::subscribe));
```
Comment (3Hren) [inline unicorn/src/service/unicorn.cpp]: ASCII art =/

File: unicorn/src/unicorn/errors.cpp [hunk 4]
```diff
@@ -100,4 +102,13 @@ auto
 make_error_code(unicorn_errors code) -> std::error_code {
     return std::error_code(static_cast<int>(code), unicorn_category());
 }
+
+namespace {
+int placeholder = []() {
```
Comment (3Hren) [inline unicorn/src/unicorn/errors.cpp]: You can do it in a place when you register the service itself.

**Comment by 3Hren** [comment]
Especially want to say "thank you" for providing test cases. This is great!

File: unicorn/include/cocaine/unicorn/writable.hpp (hunk 2 above)
Comment (antmat) [inline unicorn/include/cocaine/unicorn/writable.hpp]: yes, it is empty, so there is no profit of adding it to default message.

File: unicorn/src/service/unicorn.cpp (hunk 3 above)
Comment (antmat) [inline unicorn/src/service/unicorn.cpp]: Looks great, isn't it? :-)

File: unicorn/src/cluster/unicorn.cpp [hunk 5]
```diff
@@ -186,8 +251,8 @@ unicorn_cluster_t::on_list_update::write(unicorn::api_t::response::children_subs
 }
 
 void
-unicorn_cluster_t::on_list_update::abort(const std::error_code& rc) {
-    COCAINE_LOG_WARNING(parent->log, "failure during subscription(%i): %s, resubscribing", rc.value(), rc.message().c_str());
+unicorn_cluster_t::on_list_update::abort(const std::error_code& rc, const std::string& reason) {
+    COCAINE_LOG_WARNING(parent->log, "failure during subscription(%i): %s - %s, resubscribing", rc.value(), rc.message().c_str(), reason);
```
Comment (3Hren) [inline unicorn/src/cluster/unicorn.cpp]: Shouldn't we use `error::to_string(rc)` here?

File: unicorn/src/cluster/unicorn.cpp [hunk 6]
```diff
@@ -230,15 +293,15 @@ unicorn_cluster_t::announce() {
...
-            // Drop all ephemeral nodes, so create will succeed.
-            // It's the easiest way to drop old announces as all clients will receive a notify on delete-create.
-            auto lock = unicorn.synchronize();
-            zk.reconnect();
-        }
+    if(!endpoints.empty() && cur_endpoints != endpoints) {
+        // TODO: fix this case, if ever we would need it
+        // This can happen only if actor endpoints have changed
+        // which is not likely.
+        BOOST_ASSERT_MSG(false, "endpoints changed for locator sercice, can not comtimue, terminating" );
```
Comment (3Hren) [inline unicorn/src/cluster/unicorn.cpp]: s/comtimue/continue

File: unicorn/src/service/unicorn.cpp [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+namespace {
+class null_scope_t:
+    public api::unicorn_scope_t
+{
+public:
+    virtual void
+    close() {}
+
+    virtual
+    ~null_scope_t(){}
```
Comment (3Hren) [inline unicorn/src/service/unicorn.cpp]: Why? It's perfectly defaulted.

**Comment by 3Hren** [comment]
We can fix those small issues later.

**Pull request merged**